keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning/README.md

```diff
@@ -0,0 +1,941 @@
+-   DynamicProvisions []*v1.PersistentVolumeClaim
++ 	DynamicProvisions []*DynamicProvision
+}
+```
+
+### Get the capacity of nodes for dynamic provisioning
+
+Add `CSIStorageCapacity` to the return value of the `volumeBinder.hasEnoughCapacity` method. This returns the `DynamicProvision.Capacity` field in the case of dynamic provisioning.
+
+```diff
+- func (b *volumeBinder) hasEnoughCapacity(provisioner string, claim *v1.PersistentVolumeClaim, storageClass *storagev1.StorageClass, node *v1.Node) (bool, error) {
++ func (b *volumeBinder) hasEnoughCapacity(provisioner string, claim *v1.PersistentVolumeClaim, storageClass *storagev1.StorageClass, node *v1.Node) (bool, *storagev1.CSIStorageCapacity, error) {
+	quantity, ok := claim.Spec.Resources.Requests[v1.ResourceStorage]
+	if !ok {
+		// No capacity to check for.
+- 		return true, nil
++ 		return true, nil, nil
+	}
+
+	// Only enabled for CSI drivers which opt into it.
+	driver, err := b.csiDriverLister.Get(provisioner)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Either the provisioner is not a CSI driver or the driver does not
+			// opt into storage capacity scheduling. Either way, skip
+			// capacity checking.
+- 			return true, nil
++ 			return true, nil, nil
+		}
+- 		return false, err
++ 		return false, nil, err
+	}
+	if driver.Spec.StorageCapacity == nil || !*driver.Spec.StorageCapacity {
+- 		return true, nil
++ 		return true, nil, nil
+	}
+
+	// Look for a matching CSIStorageCapacity object(s).
+	// TODO (for beta): benchmark this and potentially introduce some kind of lookup structure (https://github.com/kubernetes/enhancements/issues/1698#issuecomment-654356718).
+	capacities, err := b.csiStorageCapacityLister.List(labels.Everything())
+	if err != nil {
+- 		return false, err
++ 		return false, nil, err
+	}
+
+  sizeInBytes := quantity.Value()
+	for _, capacity := range capacities {
+		if capacity.StorageClassName == storageClass.Name &&
+			capacitySufficient(capacity, sizeInBytes) &&
+			b.nodeHasAccess(node, capacity) {
+			// Enough capacity found.
+- 			return true, nil
++ 			return true, capacity, nil
+		}
+	}
+
+	// TODO (?): this doesn't give any information about which pools where considered and why
+	// they had to be rejected. Log that above? But that might be a lot of log output...
+	klog.V(4).InfoS("Node has no accessible CSIStorageCapacity with enough capacity for PVC",
+		"node", klog.KObj(node), "PVC", klog.KObj(claim), "size", sizeInBytes, "storageClass", klog.KObj(storageClass))
+- 	return false, nil
++ 	return false, nil, nil
+}
+```
+
+### Scoring of nodes for dynamic provisioning
+
+Add scoring of nodes for dynamic provisioning in the `Score` method of the VolumeBinding plugin. The scoring applies to every entry in `podVolumes.DynamicProvisions` where `Capacity` is not equal to `nil`.
+
+Scoring is implemented using the existing StaticBindings mechanism. Add the following to `classResources`, which will be passed to the `scorer` function:
```

Comment:
Do you think it is possible to have conflicting priorities between pre-created PVs and dynamically provisioned PVs?

The goal of https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1845-prioritization-on-volume-capacity, is to prioritize the best utilization of requested/capacity.

However, that could potentially conflict with the goals for dynamic provisioning, which could be to optimize for bin-packing vs spreading.

Should we consider evaluating the capacities separately and have an additional factor to prioritize pre-existing disks vs dynamically provisioning a disk?

Reply:
@msau42 
> Do you think it is possible to have conflicting priorities between pre-created PVs and dynamically provisioned PVs?

No, I don't. I understand from the following statement in the document[1] that dynamically provisioned PV is used if a pre-created PV is not matched. What are your specific concerns?

> When none of the static PVs the administrator created match a user's PersistentVolumeClaim, the cluster may try to dynamically provision a volume specially for the PVC.

[1] https://kubernetes.io/docs/concepts/storage/persistent-volumes/#dynamic

Reply:
Right now I am undestanding the proposal to add both the existing pre-created PV score to the dynamic provisioning score.

Say you have a node where there is both available pre-created PVs and there is a pool for dynamic provisioning.

What if the score for pre-created PVs is very high, for example, 90. And then the score for dynamic provisioning is low, like 10. If we add the two scores together, then the two scores cancel each other out and it becomes a suboptimal choice for both configurations. Potentially this leads to more fragmentation since neither scenario is being optimized.

Reply:
I'm confirming this. I will probably answer tomorrow.

Reply:
This is discussed in our team. The answer will be after next week.

Reply:
I think the issue I am thinking of is when a single storageclass can handle both dynamic + static provisioning. My preference would be for that case to not have the scores be averaged.

Reply:
@msau42 
I finally understand what you are trying to say. Thank you for your explanation. Please wait while I think of the idea for an improvement.

Reply:
@msau42 
For the moment, I will rewrite the KEP with the content of B. Please let me know if you have any feedback.

Reply:
@msau42 
I addressed it.

Reply:
With method B, you'll get the same weighted score for both dynamic and static provisioning regardless of their respective scores?  Method A seems to be more precise.

Reply:
@xing-yang 
Thank you for your comment. I'm confirming with our team. I will probably be able to reply this week.

Reply:
@xing-yang 
I'll reflect your comments and update the PR later. I answered the following question and ask a question.

> With method B, you'll get the same weighted score for both dynamic and static provisioning regardless of their respective scores?

Right. The final score of the `volumebinding` is `<the weight of volumebinding plugin> * (<static-provisioned volumes' score> + <dynamic-provisioned volumes' score>)` 

> Method A seems to be more precise.

Got it.

I realized that [my last message]([https://github.com/kubernetes/enhancements/pull/4042#discussion_r1361799115)](https://github.com/kubernetes/enhancements/pull/4042#discussion_r1361799115%29) was ambigious.
So let me clarify what I meant in the description of method A("Separate scores for static and dynamic provisioning").

Since weights exist for each scheduler plugin (correct me if I'm wrong), we should add another scheduler plugin for dynamic-provisioned volumes.
I think it's going too far.

As a compromise, I proposed the following implementation idea:

```
func (pl *VolumeBinding) Score(...) (int64, *framework.Status) {
  ...
  if len(static) != 0 {
    return static_score;  // Same value as the current method
  } else {
    return dynamic_score;  // Propose in this KEP
  }
}
```

The score will be calculated with the existing way (only static provisioning is taken into account) if at least one PVC was statically provisioned.
Otherwise, the score will be calculated from dynamic provisioning. Is this mothod acceptable?

Reply:
`static` refers to `StaticBindings []*BindingInfo`? This means if there are both statically and dynamically PVCs, it will use the static score?

Reply:
Yes, both are correct. I will revise the "[Scoring of nodes for dynamic provisioning](https://github.com/kubernetes/enhancements/blob/8298100c06164029d8b907afe5fa628df6b0fedf/keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning/README.md#scoring-of-nodes-for-dynamic-provisioning)" section in the "Design Details" to reflect this.

Reply:
I revised it.